File: Android_2018/各种实用代码封装/SPUtil.java
package cn.sharelink.leddemo.util;

import android.annotation.TargetApi;
import android.content.Context;
import android.content.SharedPreferences;
import android.os.Build;

import cn.sharelink.leddemo.App;

public class SPUtil {

    private static final String PREF_NAME = "app_pref";
    private static SPUtil spUtil;
    private static boolean sIsAtLeastGB;

    static {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.GINGERBREAD) {
            sIsAtLeastGB = true;
        }
    }

    public SPUtil(){

    }

    public static SPUtil getInstance() {
        if (spUtil == null){
            synchronized (SPUtil.class){
                if (spUtil == null){
                    spUtil = new SPUtil();
                }
            }
        }
        return spUtil;
    }

    @TargetApi(Build.VERSION_CODES.GINGERBREAD)
    private static void apply(SharedPreferences.Editor editor) {
        if (sIsAtLeastGB) {
            editor.apply();
        } else {
            editor.commit();
        }
    }

    @TargetApi(Build.VERSION_CODES.HONEYCOMB)
    public SharedPreferences getPreferences() {
        SharedPreferences pre = App.getInstance().getSharedPreferences(PREF_NAME,
                Context.MODE_MULTI_PROCESS);
        return pre;
    }

    public void set(String key, int value) {
        SharedPreferences.Editor editor = getPreferences().edit();
        editor.putInt(key, value);
        apply(editor);
    }

    public void set(String key, boolean value) {
        SharedPreferences.Editor editor = getPreferences().edit();
        editor.putBoolean(key, value);
        apply(editor);
    }

    public void set(String key, String value) {
        SharedPreferences.Editor editor = getPreferences().edit();
        editor.putString(key, value);
        apply(editor);
    }

    public boolean get(String key, boolean defValue) {
        return getPreferences().getBoolean(key, defValue);
    }

    public String get(String key, String defValue) {
        return getPreferences().getString(key, defValue);
    }

    public int get(String key, int defValue) {
        return getPreferences().getInt(key, defValue);
    }
}

File: Android_2018/aaaKotlin/SPUtil.kt
package com.li.libook.util

import android.annotation.TargetApi
import android.content.Context
import android.content.SharedPreferences
import android.os.Build
import com.example.limap.App
import com.li.libook.util.SPUtil.Companion.PREF_NAME


class SPUtil {

    private val preferences: SharedPreferences
        @TargetApi(Build.VERSION_CODES.HONEYCOMB)
        get() = App.instance.getSharedPreferences(
            PREF_NAME,
            Context.MODE_MULTI_PROCESS
        )

    operator fun set(key: String, value: Int) {
        val editor = preferences.edit()
        editor.putInt(key, value)
        apply(editor)
    }

    operator fun set(key: String, value: Boolean) {
        val editor = preferences.edit()
        editor.putBoolean(key, value)
        apply(editor)
    }

    operator fun set(key: String, value: String) {
        val editor = preferences.edit()
        editor.putString(key, value)
        apply(editor)
    }

    operator fun get(key: String, defValue: Boolean): Boolean {
        return preferences.getBoolean(key, defValue)
    }

    operator fun get(key: String, defValue: String): String? {
        return preferences.getString(key, defValue)
    }

    operator fun get(key: String, defValue: Int): Int {
        return preferences.getInt(key, defValue)
    }

    companion object {

        private const val PREF_NAME = "li_map"
        private var spUtil: SPUtil? = null
        private var sIsAtLeastGB: Boolean = false

        init {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.GINGERBREAD) {
                sIsAtLeastGB = true
            }
        }

        val instance: SPUtil
            get() {
                if (spUtil == null) {
                    synchronized(SPUtil::class.java) {
                        if (spUtil == null) {
                            spUtil = SPUtil()
                        }
                    }
                }
                return spUtil!!
            }

        @TargetApi(Build.VERSION_CODES.GINGERBREAD)
        private fun apply(editor: SharedPreferences.Editor) {
            if (sIsAtLeastGB) {
                editor.apply()
            } else {
                editor.commit()
            }
        }
    }
}
